Add macOS arm64 build

sebdanielsson opened the pull request:
Current universal build doesn't work. I added another job to build a separate macOS arm64 version. Having them separated also reduces the size of the application.

Closes #380

grulja commented:
Have you verified the new build on apple silicon? I have only old Intel based macBook.

sebdanielsson commented:
I've tested to start the build created from this PR and macOS starts it as a native application without using Rosetta.
<img width="1440" alt="Screenshot 2023-11-02 at 15 21 55" src="https://github.com/FedoraQt/MediaWriter/assets/20663065/c20a8a42-e22b-4378-94d4-12625b3e12e2">

However I wasn't able to test the functionality since both this build and the previous build from the PR before mine result in this error when flashing the USB flash drive.
Error:
@rpath/QtDBus.framework/Versions/A/QtDBus

grulja commented:
Does https://github.com/FedoraQt/MediaWriter/pull/666 make any difference?

grulja commented:
> Does #666 make any difference?

Hmm, it does, but the error shifts to QtCore instead.

grulja commented:
Alright, I fixed the issue. Can you please rebase? By that I mean to include the same change I just merged in your job?

sebdanielsson commented:
Hmm, that's strange. I rebased your fixes on top of my commits and added:
```
"$Qt6_DIR"/bin/macdeployqt build/src/app/FedoraMediaWriter.app -qmldir=src/app/qml -executable=build/src/app/FedoraMediaWriter.app/Contents/MacOS/helper -appstore-compliant
```

to the macos-arm64 build but now macOS won't open the app because it's damaged.

grulja commented:
> Hmm, that's strange. I rebased your fixes on top of my commits and added:
> 
> ```
> "$Qt6_DIR"/bin/macdeployqt build/src/app/FedoraMediaWriter.app -qmldir=src/app/qml -executable=build/src/app/FedoraMediaWriter.app/Contents/MacOS/helper -appstore-compliant
> ```
> 
> to the macos-arm64 build but now macOS won't open the app because it's damaged.

I don't think that any change I did could damage the build/bundle somehow, it just added yet another binary to work with in the `macdeployqt` to fix the libraries and paths, but nothing really fundamental. Does the non-arm64 build work for you? Or both builds are now broken? It at least worked for me yesterday.

sebdanielsson commented:
> > Hmm, that's strange. I rebased your fixes on top of my commits and added:
> > ```
> > "$Qt6_DIR"/bin/macdeployqt build/src/app/FedoraMediaWriter.app -qmldir=src/app/qml -executable=build/src/app/FedoraMediaWriter.app/Contents/MacOS/helper -appstore-compliant
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to the macos-arm64 build but now macOS won't open the app because it's damaged.
> 
> I don't think that any change I did could damage the build/bundle somehow, it just added yet another binary to work with in the `macdeployqt` to fix the libraries and paths, but nothing really fundamental. Does the non-arm64 build work for you? Or both builds are now broken? It at least worked for me yesterday.

The x86_64 version still works just like before, only the arm64 version is reported as broken.

grulja commented:
That's weird. Any chance you can run it in a terminal to get some logs why it reports it's broken? Or it doesn't even mount?

sebdanielsson commented:
Starting the app from the terminal doesn't give me any output. I started the macOS Console.app and started the app and got lot's of output. Here's the output filtered by `MediaWriter`:

<details>
<summary>MediaWriter Console Output</summary>

```
default	10:21:00.583832+0100	lsd	org.fedoraproject.MediaWriter/Unknown Persona:5:5:1:1:Building bundle record for app
default	10:21:00.584050+0100	lsd	org.fedoraproject.MediaWriter/Unknown Persona:5:5:2:1:Built bundle record for app
default	10:21:00.592417+0100	lsd	org.fedoraproject.MediaWriter/Unknown Persona:5:4:2:1:_LSServerRegisterItemInfo result = 0
default	10:21:00.602942+0100	runningboardd	Launch request for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>[0] is using uid 501 (divined from auid 501 euid 501)
default	10:21:00.602998+0100	runningboardd	Acquiring assertion targeting app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)> from originator [osservice<com.apple.Finder(501)>:1601] with description <RBSAssertionDescriptor| "LS launch org.fedoraproject.MediaWriter" ID:604-1601-452149 target:app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)> attributes:[
	<RBSDomainAttribute| domain:"com.apple.launchservicesd" name:"LaunchRoleUserInteractive" sourceEnvironment:"(null)">
	]>
default	10:21:00.603056+0100	runningboardd	Assertion 604-1601-452149 (target:app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>) will be created as active
default	10:21:00.603283+0100	runningboardd	Executing launch request for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)> (LS launch org.fedoraproject.MediaWriter)
default	10:21:00.603552+0100	runningboardd	Checking PreventLaunch: global:0 exPath:/Applications/FedoraMediaWriter.app/Contents/MacOS/FedoraMediaWriter predicates:(null) allow:(null)
default	10:21:00.603929+0100	runningboardd	Creating and launching job for: app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>
default	10:21:00.604175+0100	runningboardd	_mutateContextIfNeeded called for org.fedoraproject.MediaWriter
default	10:21:00.604722+0100	runningboardd	app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>: -[RBPersonaManager personaForIdentity:context:personaUID:personaUniqueString:] required 0.000000 ms (wallclock); resolved to {4294967295, (null)}
default	10:21:00.605015+0100	runningboardd	'app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>' Constructed job description:
<dictionary: 0x14684e6c0> { count = 23, transaction: 0, voucher = 0x0, contents =
	"Platform" => <int64: 0x9799cfffba4ac997>: 1
	"ProcessType" => <string: 0x141e456d0> { length = 3, contents = "App" }
	"EnableTransactions" => <bool: 0x1eaa617a0>: false
	"_ManagedBy" => <string: 0x141e4bcf0> { length = 22, contents = "com.apple.runningboard" }
	"CFBundleIdentifier" => <string: 0x141e20b80> { length = 29, contents = "org.fedoraproject.MediaWriter" }
	"_ResourceCoalition" => <string: 0x141e54f70> { length = 71, contents = "app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>" }
	"_DisablePointerAuth" => <bool: 0x1eaa61780>: true
	"ThrottleInterval" => <int64: 0x9799cffc45b53667>: 2147483647
	"MachServices" => <dictionary: 0x141e2ee70> { count = 0, transaction: 0, voucher = 0x0, contents =
	}
	"EnablePressuredExit" => <bool: 0x1eaa617a0>: false
	"LimitLoadToSessionType" => <array: 0x141e43b60> { count = 2, capacity = 8, contents =
		0: <string: 0x141e6c500> { length = 4, contents = "Aqua" }
		1: <string: 0x141e39ac0> { length = 11, contents = "LoginWindow" }
	}
	"InitialTaskRole" => <int64: 0x9799cfffba4ac98f>: 2
	"EnvironmentVariables" => <dictionary: 0x141e450e0> { count = 12, transaction: 0, voucher = 0x0, contents =
		"__CF_USER_TEXT_ENCODING" => <string: 0x141e7f1f0> { length = 9, contents = "0x1F5:0:0" }
		"TMPDIR" => <string: 0x141e4a4c0> { length = 49, contents = "/var/folders/dj/g6nlcz0d12sfdbmdfszv8br00000gn/T/" }
		"SHELL" => <string: 0x141e10f20> { length = 8, contents = "/bin/zsh" }
		"HOME" => <string: 0x141e94290> { length = 16, contents = "/Users/sebastian" }
		"SSH_AUTH_SOCK" => <string: 0x141e40c20> { length = 51, contents = "/private/tmp/com.apple.launchd.CaYk8MHiEo/Listeners" }
		"LOGNAME" => <string: 0x141eb7d40> { length = 9, contents = "sebastian" }
		"PATH" => <string: 0x141e83600> { length = 518, contents = "/usr/local/bin:/opt/homebrew/opt/node@18/bin:/opt/homebrew/opt/openjdk/bin:/Users/sebastian/Library/Python/3.10/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/Applications/Little Snitch.app/Contents/Components:/usr/local/share/dotnet:~/.dotnet/tools:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Applications/Wireshark.app/Contents/MacOS:/Users/sebastian/.fig/bin:/Users/sebastian/.local/bin" }
		"XPC_SERVICE_NAME" => <string: 0x141e6e700> { length = 16, contents = "com.apple.Finder" }
		"__CFBundleIdentifier" => <string: 0x141e372a0> { length = 29, contents = "org.fedoraproject.MediaWriter" }
		"COMMAND_MODE" => <string: 0x141e91c90> { length = 8, contents = "unix2003" }
		"USER" => <string: 0x141e7c000> { length = 9, contents = "sebastian" }
		"XPC_FLAGS" => <string: 0x141e6b0b0> { length = 3, contents = "0x0" }
	}
	"_AdditionalProperties" => <dictionary: 0x14683f840> { count = 1, transaction: 0, voucher = 0x0, contents =
		"RunningBoard" => <dictionary: 0x1468634e0> { count = 4, transaction: 0, voucher = 0x0, contents =
			"TMPDIR" => <string: 0x1468b8690> { length = 49, contents = "/var/folders/dj/g6nlcz0d12sfdbmdfszv8br00000gn/T/" }
			"HOME" => <string: 0x14684dfb0> { length = 16, contents = "/Users/sebastian" }
			"RunningBoardLaunchedIdentity" => <dictionary: 0x1468fab40> { count = 5, transaction: 0, voucher = 0x0, contents =
				"AJL" => <string: 0x146883b60> { length = 61, contents = "application.org.fedoraproject.MediaWriter.185225543.185225775" }
				"TYPE" => <int64: 0x9799cfffba4ac997>: 1
				"AUID" => <uint64: 0x9719cfffba4ac637>: 501
				"EAI" => <string: 0x14684e850> { length = 29, contents = "org.fedoraproject.MediaWriter" }
				"PLAT" => <uint64: 0x9719cfffba4ac997>: 1
			}
			"RunningBoardLaunched" => <bool: 0x1eaa61780>: true
		}
	}
	"ExitTimeOut" => <int64: 0x9799cfffba4ac997>: 1
	"Label" => <string: 0x141e14810> { length = 61, contents = "application.org.fedoraproject.MediaWriter.185225543.185225775" }
	"WaitForDebugger" => <bool: 0x1eaa61780>: true
	"MaterializeDatalessFiles" => <bool: 0x1eaa61780>: true
	"WorkingDirectory" => <string: 0x141e5af40> { length = 1, contents = "/" }
	"_LaunchType" => <int64: 0x9799cfffba4ac987>: 3
	"AbandonProcessGroup" => <bool: 0x1eaa61780>: true
	"ProgramArguments" => <array: 0x141e5a590> { count = 1, capacity = 8, contents =
		0: <string: 0x141e71ea0> { length = 68, contents = "/Applications/FedoraMediaWriter.app/Contents/MacOS/FedoraMediaWriter" }
	}
	"Program" => <string: 0x141e5b580> { length = 68, contents = "/Applications/FedoraMediaWriter.app/Contents/MacOS/FedoraMediaWriter" }
}
default	10:21:00.615334+0100	runningboardd	_bundleMatchesProcessWithExecutablePath using realpath and comparing /Applications/FedoraMediaWriter.app/Contents/MacOS/FedoraMediaWriter and /Applications/FedoraMediaWriter.app/Contents/MacOS
default	10:21:00.615534+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Memory Limits: active 0 inactive 0
 <private>
default	10:21:00.615549+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] This process will not be managed.
default	10:21:00.615561+0100	runningboardd	Now tracking process: [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:00.615854+0100	runningboardd	Calculated state for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>: running-active (role: UserInteractive)
default	10:21:00.616661+0100	runningboardd	Using default underlying assertion for app: [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:00.617177+0100	runningboardd	Acquiring assertion targeting [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] from originator [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] with description <RBSAssertionDescriptor| "RB Underlying Assertion" ID:604-604-452150 target:16036 attributes:[
	<RBSDomainAttribute| domain:"com.apple.underlying" name:"defaultUnderlyingAppAssertion" sourceEnvironment:"(null)">,
	<RBSAcquisitionCompletionAttribute| policy:AfterApplication>
	]>
default	10:21:00.617402+0100	runningboardd	Assertion 604-604-452150 (target:[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]) will be created as active
default	10:21:00.617695+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring jetsam update because this process is not memory-managed
default	10:21:00.617856+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring suspend because this process is not lifecycle managed
default	10:21:00.618267+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] reported to RB as running
default	10:21:00.618145+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Set darwin role to: UserInteractive
default	10:21:00.618966+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring GPU update because this process is not GPU managed
default	10:21:00.619637+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring memory limit update because this process is not memory-managed
default	10:21:00.619905+0100	runningboardd	Calculated state for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>: running-active (role: UserInteractive)
default	10:21:00.621500+0100	gamepolicyd	Hit the server for a process handle 1c8845c500003ea4 that resolved to: [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:00.621540+0100	gamepolicyd	Received state update for 16036 (app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>, running-active-NotVisible
default	10:21:00.623260+0100	runningboardd	Acquiring assertion targeting [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] from originator [osservice<com.apple.coreservices.launchservicesd>:568] with description <RBSAssertionDescriptor| "foregroundApp:16036" ID:604-568-452151 target:16036 attributes:[
	<RBSDomainAttribute| domain:"com.apple.launchservicesd" name:"RoleUserInteractiveNonFocal" sourceEnvironment:"(null)">
	]>
default	10:21:00.623429+0100	runningboardd	Assertion 604-568-452151 (target:[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]) will be created as active
default	10:21:00.624593+0100	runningboardd	Calculated state for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>: running-active (role: UserInteractive)
default	10:21:00.627262+0100	runningboardd	Invalidating assertion 604-1601-452149 (target:app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>) from originator [osservice<com.apple.Finder(501)>:1601]
default	10:21:00.630646+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring jetsam update because this process is not memory-managed
default	10:21:00.630990+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring suspend because this process is not lifecycle managed
default	10:21:00.631145+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring GPU update because this process is not GPU managed
default	10:21:00.631469+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring memory limit update because this process is not memory-managed
default	10:21:00.631719+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring jetsam update because this process is not memory-managed
default	10:21:00.632198+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring suspend because this process is not lifecycle managed
default	10:21:00.632277+0100	runningboardd	Successfully acquired underlying assertion for [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:00.632322+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring GPU update because this process is not GPU managed
default	10:21:00.632605+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring memory limit update because this process is not memory-managed
default	10:21:00.625838+0100	distnoted	register name: com.apple.sharedfilelist.change object: com.apple.LSSharedFileList.ApplicationRecentDocuments/org.fedoraproject.mediawriter token: 1380000005e pid: 1599
default	10:21:00.643215+0100	runningboardd	Calculated state for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>: running-active (role: UserInteractiveNonFocal)
default	10:21:00.643388+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring jetsam update because this process is not memory-managed
default	10:21:00.643522+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring suspend because this process is not lifecycle managed
default	10:21:00.643584+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Set darwin role to: UserInteractiveNonFocal
default	10:21:00.643620+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring GPU update because this process is not GPU managed
default	10:21:00.643711+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] Ignoring memory limit update because this process is not memory-managed
default	10:21:00.649528+0100	loginwindow	-[PersistentAppsSupport applicationReady:] | App: FedoraMediaWriter, ready, updating active tracking timer
default	10:21:00.649564+0100	loginwindow	-[ApplicationManager checkInAppContext:eventData:] | ApplicationManager: Checked in app : FedoraMediaWriter
default	10:21:00.651648+0100	gamepolicyd	Received state update for 16036 (app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>, running-active-NotVisible
default	10:21:00.723470+0100	gamepolicyd	Received state update for 16036 (app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>, running-active-NotVisible
default	10:21:00.736493+0100	dmd	Received xpc stream event (distributed notification matching) with name: com.apple.LaunchServices.applicationRegistered user info: {
    CFBundleDisplayName = "Fedora Media Writer";
    bundleIDs =     (
        "org.fedoraproject.MediaWriter"
    );
    isPlaceholder = 0;
}
default	10:21:02.223534+0100	tccd	AUTHREQ_ATTRIBUTION: msgID=703.123, attribution={accessing={TCCDProcess: identifier=FedoraMediaWriter, pid=16036, auid=501, euid=501, binary_path=/Applications/FedoraMediaWriter.app/Contents/MacOS/FedoraMediaWriter}, requesting={TCCDProcess: identifier=com.apple.syspolicyd, pid=703, auid=0, euid=0, binary_path=/usr/libexec/syspolicyd}, },
default	10:21:02.237401+0100	tccd	AUTHREQ_SUBJECT: msgID=703.123, subject=org.fedoraproject.MediaWriter,
default	10:21:02.239571+0100	tccd	-[TCCDAccessIdentity staticCode]: static code for: identifier org.fedoraproject.MediaWriter, type: 0: 0x124f9f120 at /Applications/FedoraMediaWriter.app
default	10:21:02.243864+0100	kernel	ASP: Security policy would not allow process: 16036, /Applications/FedoraMediaWriter.app/Contents/MacOS/FedoraMediaWriter
default	10:21:02.244656+0100	runningboardd	[app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036] termination reported by launchd (9, 8, 9)
default	10:21:02.244718+0100	runningboardd	Removing process: [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:02.244960+0100	runningboardd	Removing launch job for: [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:02.245208+0100	runningboardd	Removed job for [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:02.245242+0100	runningboardd	Removing assertions for terminated process: [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:02.254071+0100	runningboardd	Calculated state for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>: none (role: None)
default	10:21:02.254292+0100	runningboardd	Calculated state for app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>: none (role: None)
default	10:21:02.256031+0100	gamepolicyd	Received state update for 16036 (app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>, none-NotVisible
default	10:21:02.256376+0100	launchservicesd	Hit the server for a process handle 1c8845c500003ea4 that resolved to: [app<application.org.fedoraproject.MediaWriter.185225543.185225775(501)>:16036]
default	10:21:02.268363+0100	loginwindow	-[PersistentAppsSupport applicationQuit:] | for app:FedoraMediaWriter, _appTrackingState = 2
default	10:21:02.268400+0100	loginwindow	-[PersistentAppsSupport applicationQuit:] | App: FedoraMediaWriter, quit, updating active tracking timer
```
</details>

This line stands out to me. I tried to disable Gatekeeper, but still got the same error.

```
default	10:21:02.243864+0100	kernel	ASP: Security policy would not allow process: 16036, /Applications/FedoraMediaWriter.app/Contents/MacOS/FedoraMediaWriter
```

sebdanielsson commented:
Whoa!

Actually disabling Gatekeeper and re-enabling it did the trick. It's flashing my USB stick now!

```
sudo spctl --master-disable
spctl --status
sudo spctl --master-enable
spctl --status
```

<img width="1104" alt="Screenshot 2023-11-03 at 10 30 39" src="https://github.com/FedoraQt/MediaWriter/assets/20663065/473911ad-37c5-4647-ae40-78c045e1c6c2">

grulja reviewed (approved):
Thanks!

sebdanielsson commented:
Thank you!

One difference I noticed for builds using the new CI system is that the new app filename doesn't contain spaces. Is this intentional? Many users will probably end up with two versions if it's not the same.

grulja commented:
> Thank you!
> 
> One difference I noticed for builds using the new CI system is that the new app filename doesn't contain spaces. Is this intentional? Many users will probably end up with two versions if it's not the same.

Yes, I changed it to simplify it and avoid potential issues when I omit to escape spaces or things like that in the build script. I don't think we have that many users or that many users who keep using FMW frequently on MacOS so hopefully it won't be a big deal.
